compiler/modules/front/codeGen.bal

```diff
@@ -716,22 +744,67 @@ function codeGenIfElseStmt(CodeGenContext cx, bir:BasicBlock startBlock, Environ
                 elseContBlock.insns.push(branch);
                 assignments.push(...elseAssignments);
             }
-            return { block: contBlock, assignments };
+            StmtNarrowing[] narrowings = ifNarrowings;
+            foreach var n in elseNarrowings {
+                broadenStmtNarrowings(narrowings, n);
+            }
+            return { block: contBlock, assignments, narrowings };
         }
     }
 }
 
-function codeGenIfElseNarrowing(CodeGenContext cx, bir:BasicBlock bb, Environment env, Narrowing narrowing, boolean condition, Position pos) returns Environment {
+// If already exists in the list, union, else append
+function broadenStmtNarrowings(StmtNarrowing[] list, StmtNarrowing annex) {
+    foreach int i in 0..<list.length() {
+        StmtNarrowing existing = list[i];
+        if existing.binding.name == annex.binding.name {
+            list[i] = {
+                basis: { or: [annex.basis, existing.basis] },
+                binding: annex.binding,
+                ifCompletesNormally: t:union(annex.ifCompletesNormally, existing.ifCompletesNormally)
+            };
+            return;
+        }
+    }
+    list.push(annex);
+}
+
+// If already exists in the list, replace, else append
+function tightenStmtNarrowings(StmtNarrowing[] list, StmtNarrowing annex) {
```

Comment:
We don't use broaden/tighten as a pair anywhere else. I would expect that the semantics were union and intersection and we could name accordingly.